File: react-native/src/models/Layout.ts
import { ArrangementStrategy, DimensionUnit, SpecialDimension } from './enums';

export interface Dimension {
  value: number;
  unit: DimensionUnit;
  special?: SpecialDimension | null;
}

export interface Offset {
  x: number;
  y: number;
  unit: DimensionUnit;
}

export interface Spacing {
  all?: number;
  horizontal?: number;
  vertical?: number;
  top?: number;
  bottom?: number;
  left?: number;
  right?: number;
  unit?: DimensionUnit;
}

export interface ChildArrangement {
  strategy: ArrangementStrategy;
  spacing?: number;
  spacingUnit?: DimensionUnit;
}

export interface Layout {
  width?: Dimension;
  height?: Dimension;
  aspectRatio?: number;
  offset?: Offset;
  padding?: Spacing;
  arrangement?: ChildArrangement;
}

export function resolveSpacingTop(s: Spacing): number {
  return s.top ?? s.vertical ?? s.all ?? 0;
}

export function resolveSpacingBottom(s: Spacing): number {
  return s.bottom ?? s.vertical ?? s.all ?? 0;
}

export function resolveSpacingLeft(s: Spacing): number {
  return s.left ?? s.horizontal ?? s.all ?? 0;
}

export function resolveSpacingRight(s: Spacing): number {
  return s.right ?? s.horizontal ?? s.all ?? 0;
}

export function parseDimension(raw: unknown): Dimension | undefined {
  if (raw == null) return undefined;
  if (typeof raw === 'number') {
    return { value: raw, unit: 'dp' };
  }
  if (typeof raw === 'string') {
    if (raw === 'wrap_content') return { value: 0, unit: 'dp', special: 'wrap_content' };
    if (raw === 'match_parent') return { value: 0, unit: 'dp', special: 'match_parent' };
    if (raw.endsWith('%')) {
      const val = parseFloat(raw);
      if (!isNaN(val)) return { value: val, unit: 'percent' };
    }
    const num = parseFloat(raw);
    if (!isNaN(num)) return { value: num, unit: 'dp' };
    return undefined;
  }
  if (typeof raw === 'object' && raw !== null) {
    const obj = raw as Record<string, unknown>;
    if (obj.special === 'wrap_content') return { value: 0, unit: 'dp', special: 'wrap_content' };
    if (obj.special === 'match_parent') return { value: 0, unit: 'dp', special: 'match_parent' };
    const value = typeof obj.value === 'number' ? obj.value : parseFloat(String(obj.value ?? 0));
    const unit = (obj.unit as DimensionUnit) ?? 'dp';
    const special = obj.special != null ? (obj.special as SpecialDimension) : null;
    return { value: isNaN(value) ? 0 : value, unit, special: special ?? null };
  }
  return undefined;
}

File: flutter/lib/src/models/layout.dart
import 'enums.dart';

class Dimension {
  final double value;
  final DimensionUnit unit;
  final SpecialDimension? special;

  const Dimension({
    this.value = 0,
    this.unit = DimensionUnit.dp,
    this.special,
  });

  static const wrapContent =
      Dimension(value: 0, unit: DimensionUnit.dp, special: SpecialDimension.wrapContent);
  static const matchParent =
      Dimension(value: 0, unit: DimensionUnit.dp, special: SpecialDimension.matchParent);

  factory Dimension.fromJson(Map<String, dynamic> json) {
    final special = SpecialDimension.fromJson(json['special'] as String?);
    final unit = DimensionUnit.fromJson(json['unit'] as String? ?? 'dp');
    return Dimension(
      value: (json['value'] as num?)?.toDouble() ?? 0,
      unit: unit,
      special: special,
    );
  }

  // borderRadius accepts raw number (dp) or {"value": N, "unit": "percent"}
  factory Dimension.fromJsonFlexible(dynamic json) {
    if (json == null) return const Dimension();
    if (json is num) {
      return Dimension(value: json.toDouble(), unit: DimensionUnit.dp);
    }
    if (json is Map<String, dynamic>) {
      return Dimension.fromJson(json);
    }
    return const Dimension();
  }

  @override
  String toString() => 'Dimension(value: $value, unit: $unit, special: $special)';
}

class NDOffset {
  final double x;
  final double y;
  final DimensionUnit unit;

  const NDOffset({this.x = 0, this.y = 0, this.unit = DimensionUnit.dp});

  static const zero = NDOffset();

  factory NDOffset.fromJson(Map<String, dynamic> json) => NDOffset(
        x: (json['x'] as num?)?.toDouble() ?? 0,
        y: (json['y'] as num?)?.toDouble() ?? 0,
        unit: DimensionUnit.fromJson(json['unit'] as String? ?? 'dp'),
      );
}

class Spacing {
  final double? all;
  final double? horizontal;
  final double? vertical;
  final double? top;
  final double? bottom;
  final double? left;
  final double? right;
  final DimensionUnit unit;

  const Spacing({
    this.all,
    this.horizontal,
    this.vertical,
    this.top,
    this.bottom,
    this.left,
    this.right,
    this.unit = DimensionUnit.dp,
  });

  double resolveTop() => top ?? vertical ?? all ?? 0.0;
  double resolveBottom() => bottom ?? vertical ?? all ?? 0.0;
  double resolveLeft() => left ?? horizontal ?? all ?? 0.0;
  double resolveRight() => right ?? horizontal ?? all ?? 0.0;

  factory Spacing.fromJson(Map<String, dynamic> json) => Spacing(
        all: (json['all'] as num?)?.toDouble(),
        horizontal: (json['horizontal'] as num?)?.toDouble(),
        vertical: (json['vertical'] as num?)?.toDouble(),
        top: (json['top'] as num?)?.toDouble(),
        bottom: (json['bottom'] as num?)?.toDouble(),
        left: (json['left'] as num?)?.toDouble(),
        right: (json['right'] as num?)?.toDouble(),
        unit: DimensionUnit.fromJson(json['unit'] as String? ?? 'dp'),
      );
}

class ChildArrangement {
  final double? spacing;
  final DimensionUnit spacingUnit;
  final ArrangementStrategy strategy;

  const ChildArrangement({
    this.spacing,
    this.spacingUnit = DimensionUnit.dp,
    this.strategy = ArrangementStrategy.spaced,
  });

  static const defaultArrangement =
      ChildArrangement(spacing: 0, strategy: ArrangementStrategy.spaced);

  factory ChildArrangement.fromJson(Map<String, dynamic> json) => ChildArrangement(
        spacing: (json['spacing'] as num?)?.toDouble(),
        spacingUnit: DimensionUnit.fromJson(json['spacingUnit'] as String? ?? 'dp'),
        strategy: ArrangementStrategy.fromJson(json['strategy'] as String? ?? 'spaced'),
      );
}

class Layout {
  final Dimension? width;
  final Dimension? height;
  final double? aspectRatio;
  final NDOffset? offset;
  final Spacing? padding;
  final ChildArrangement? arrangement;

  const Layout({
    this.width,
    this.height,
    this.aspectRatio,
    this.offset,
    this.padding,
    this.arrangement,
  });

  factory Layout.fromJson(Map<String, dynamic> json) => Layout(
        width:
            json['width'] != null ? Dimension.fromJson(json['width'] as Map<String, dynamic>) : null,
        height: json['height'] != null
            ? Dimension.fromJson(json['height'] as Map<String, dynamic>)
            : null,
        aspectRatio: (json['aspectRatio'] as num?)?.toDouble(),
        offset: json['offset'] != null
            ? NDOffset.fromJson(json['offset'] as Map<String, dynamic>)
            : null,
        padding: json['padding'] != null
            ? Spacing.fromJson(json['padding'] as Map<String, dynamic>)
            : null,
        arrangement: json['arrangement'] != null
            ? ChildArrangement.fromJson(json['arrangement'] as Map<String, dynamic>)
            : null,
      );
}
